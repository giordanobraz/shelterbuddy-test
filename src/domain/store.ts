import { configureStore, createSlice } from "@reduxjs/toolkit";
import { Data } from "../utils/createData.util";

interface IAnimalsStoreData {
  list: Data[];
}

const initialState: IAnimalsStoreData = {
  list: [],
};

const animalsDataSlice = createSlice({
  name: "animals",
  initialState,
  reducers: {
    addAnimal(state, action) {
      return {
        ...state,
        list: action.payload,
      };
    },
  },
});

export const { addAnimal } = animalsDataSlice.actions;
export const getAnimalsList = (state: any) => state.list;

const store = configureStore({
  reducer: animalsDataSlice.reducer,
});

export default store;
