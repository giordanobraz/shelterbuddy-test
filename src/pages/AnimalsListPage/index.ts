import { AnimalsListPage } from "./AnimalsListPage";

export { AnimalsListPage };
