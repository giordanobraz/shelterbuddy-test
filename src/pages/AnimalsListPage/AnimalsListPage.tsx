import {
  Box,
  BoxProps,
  Grid,
  Typography,
  styled,
  useMediaQuery,
} from "@mui/material";
import { Data, createData } from "../../utils/createData.util";
import { addAnimal, getAnimalsList } from "../../domain/store";
import { useDispatch, useSelector } from "react-redux";
import { useEffect, useState } from "react";

import { List } from "../../components/List/List";
import { Search } from "../../components/Search/Search";
import { Table } from "../../components/Table";
import { fetchAnimalsData } from "../../services/axios";

const CustomBadge = styled(Box)<BoxProps>({
  color: "#fff",
  backgroundColor: "#f47500",
  borderRadius: 100,
  padding: "4px 8px",
  fontSize: 13,
  fontWeight: "bold",
  margin: "0 8px",
});

export function AnimalsListPage() {
  const dispatch = useDispatch();
  const animalData = useSelector(getAnimalsList);
  const [searchValue, setSearchValue] = useState("");
  const mediaQuery = useMediaQuery("(max-width:810px)");

  useEffect(() => {
    if (animalData.length === 0) {
      fetchAnimalsData().then((response) => {
        const animalsList = response?.Data;
        let createRows: Data[] = [];

        animalsList.forEach((animal: any) => {
          createRows.push(
            createData(
              animal.Id,
              animal.Name ? animal.Name : "z",
              animal.Type.Name,
              animal.Breed.Primary.Name,
              animal.Sex.Name,
              animal.Features.PrimaryColour
            )
          );
        });

        dispatch(addAnimal(createRows));
      });
    }
  }, [dispatch, animalData]);

  const filteredAnimalList =
    searchValue.length < 1
      ? animalData
      : animalData.filter((animal: Data) => {
          return animal.name.toLowerCase().includes(searchValue.toLowerCase());
        });

  return (
    <Box px={{ xs: 2, sm: 6, lg: 12 }} pt={12}>
      <Box bgcolor="white" p={4} borderRadius={3} width="100%">
        <Grid container spacing={3}>
          <Grid item xs={12} md={7} alignSelf="center">
            <Box display="flex" alignItems="center" alignSelf="center">
              <Typography variant="h1" fontWeight={600} fontSize={22}>
                Your Animals
              </Typography>
              <CustomBadge component="span">{animalData.length}</CustomBadge>
            </Box>
          </Grid>
          <Grid item xs={12} md={5}>
            <Search searchValue={searchValue} setSearchValue={setSearchValue} />
          </Grid>
        </Grid>
        <Box display="flex" mt={2}>
          <Box flexGrow={1} overflow={mediaQuery ? "scroll" : "hidden"}>
            {mediaQuery ? (
              <List filteredAnimalList={filteredAnimalList} />
            ) : (
              <Table filteredAnimalList={filteredAnimalList} />
            )}
          </Box>
        </Box>
      </Box>
    </Box>
  );
}
