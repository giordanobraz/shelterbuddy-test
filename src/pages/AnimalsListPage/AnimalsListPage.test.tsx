import { render, screen } from "@testing-library/react";

import { AnimalsListPage } from "./AnimalsListPage";
import { Provider } from "react-redux";
import store from "../../domain/store";
import userEvent from "@testing-library/user-event";

describe("AnimalListPage", () => {
  it("should render", () => {
    const { container } = getRenderer();
    expect(container).toMatchSnapshot();
  });

  it("should search and find an animal", async () => {
    render(
      <Provider store={store}>
        <AnimalsListPage />
      </Provider>
    );

    const searchInput = screen.getByPlaceholderText("Search an animal by name");
    userEvent.type(searchInput, "big");

    expect(await screen.findByText("Big Sean")).toBeInTheDocument();
  });

  it("should show a message if it doesn't find an animal", async () => {
    render(
      <Provider store={store}>
        <AnimalsListPage />
      </Provider>
    );

    const searchInput = screen.getByPlaceholderText("Search an animal by name");
    userEvent.type(searchInput, "Donald");
    expect(await screen.findByText("No results found.")).toBeInTheDocument();
  });
});

function getRenderer() {
  return render(
    <Provider store={store}>
      <AnimalsListPage />
    </Provider>
  );
}
