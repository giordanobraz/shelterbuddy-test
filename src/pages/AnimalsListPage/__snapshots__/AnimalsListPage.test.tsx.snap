// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimalListPage should render 1`] = `
<div>
  <div
    class="MuiBox-root css-9owtmb"
  >
    <div
      class="MuiBox-root css-8g4y2h"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-7 css-1jge30o-MuiGrid-root"
        >
          <div
            class="MuiBox-root css-13bort8"
          >
            <h1
              class="MuiTypography-root MuiTypography-h1 css-1asadsr-MuiTypography-root"
            >
              Your Animals
            </h1>
            <span
              class="MuiBox-root css-hsggji"
            >
              0
            </span>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-5 css-urj5pv-MuiGrid-root"
        >
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-adornedStart css-13jorxd-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart css-ittuaa-MuiInputAdornment-root"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchOutlinedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
            </div>
            <input
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
              placeholder="Search an animal by name"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-hdw1oc"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-mifb2i"
      >
        <div
          class="MuiBox-root css-ndu8ol"
        >
          <div
            class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
          >
            <table
              class="MuiTable-root css-x9u59m-MuiTable-root"
            >
              <thead
                class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1sjvjxo-MuiTableCell-root"
                    scope="col"
                  >
                    <span
                      class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                      role="button"
                      tabindex="0"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                        data-testid="ArrowDownwardIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                        />
                      </svg>
                    </span>
                  </th>
                  <th
                    aria-sort="ascending"
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1sjvjxo-MuiTableCell-root"
                    scope="col"
                  >
                    <span
                      class="MuiButtonBase-root MuiTableSortLabel-root Mui-active css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                      role="button"
                      tabindex="0"
                    >
                      Name
                      <span
                        class="MuiBox-root css-trfuzi"
                      >
                        sorted ascending
                      </span>
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                        data-testid="ArrowDownwardIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                        />
                      </svg>
                    </span>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1sjvjxo-MuiTableCell-root"
                    scope="col"
                  >
                    <span
                      class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                      role="button"
                      tabindex="0"
                    >
                      Type
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                        data-testid="ArrowDownwardIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                        />
                      </svg>
                    </span>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1sjvjxo-MuiTableCell-root"
                    scope="col"
                  >
                    <span
                      class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                      role="button"
                      tabindex="0"
                    >
                      Breed
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                        data-testid="ArrowDownwardIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                        />
                      </svg>
                    </span>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1sjvjxo-MuiTableCell-root"
                    scope="col"
                  >
                    <span
                      class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                      role="button"
                      tabindex="0"
                    >
                      Gender
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                        data-testid="ArrowDownwardIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                        />
                      </svg>
                    </span>
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1sjvjxo-MuiTableCell-root"
                    scope="col"
                  >
                    <span
                      class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
                      role="button"
                      tabindex="0"
                    >
                      Color
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
                        data-testid="ArrowDownwardIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                        />
                      </svg>
                    </span>
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
              >
                <tr
                  class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
                >
                  <td
                    class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      No results found.
                    </p>
                  </td>
                </tr>
              </tbody>
            </table>
            <div
              class="css-1u96uvg-MuiStack-root"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
